src/test/groovy/com/cinnober/gradle/semver_git/SemverGitPluginFunctionalTest.groovy

```diff
@@ -82,44 +89,246 @@ class SemverGitPluginFunctionalTest extends Specification {
+        where:
+        config << configs(snapshotSuffix: "SNAPSHOT<dirty>")
+    }
+
+    def "Dirty untagged commit with default config"() {
+        when:
+        setupPluginWithClasspath(config)
+        init()
+        commit()
+        tag("1.0.0")
+        commit()
+        touchBuildFile()
+        BuildResult result = gradleBuild("showVersion")
+
+        then:
+        result.getOutput().contains("Version: 1.1.0-SNAPSHOT")
+
+        where:
+        config << configs()
+    }
+
+    def "Dirty untagged commit with snapshotSuffix with <dirty>"() {
+        when:
+        setupPluginWithClasspath(config)
+        init()
+        commit()
+        tag("1.0.0")
+        commit()
+        touchBuildFile()
+        BuildResult result = gradleBuild("showVersion")
+
+        then:
+        result.getOutput().contains("Version: 1.1.0-SNAPSHOT-dirty")
+
+        where:
+        config << configs(snapshotSuffix: "SNAPSHOT<dirty>")
+    }
+
+    def "Custom nextVersion"() {
+        when:
+        setupPluginWithClasspath(config)
+        init()
+        commit()
+        tag("1.0.0")
+        commit()
+        BuildResult result = gradleBuild("showVersion")
+
+        then:
+        result.getOutput().contains("Version: 2.0.0-SNAPSHOT")
+
+        where:
+        config << configs(nextVersion: "major")
+    }
+
+    def "Custom snapshotSuffix"() {
+        when:
+        setupPluginWithClasspath(config)
+        init()
+        commit()
+        tag("1.0.0")
+        commit()
+        BuildResult result = gradleBuild("showVersion")
+
+        then:
+        result.getOutput().contains("Version: 1.1.0-volatile")
+
+        where:
+        config << configs(snapshotSuffix: "volatile")
+    }
+
+    def "Custom dirtyMarker"() {
+        when:
+        setupPluginWithClasspath(config)
+        init()
+        commit()
+        tag("1.0.0")
+        commit()
+        touchBuildFile()
+        BuildResult result = gradleBuild("showVersion")
+
+        then:
+        result.getOutput().contains("Version: 1.1.0-SNAPSHOT_MODIFIED")
+
+        where:
+        config << configs(snapshotSuffix: "SNAPSHOT<dirty>", dirtyMarker: "_MODIFIED")
+    }
+
+    def "Custom gitDescribeArgs"() {
+        when:
+        setupPluginWithClasspath(config)
+        init()
+        String shortCommit = commit()
+        tag("1.0.0")
+        BuildResult result = gradleBuild("showVersion")
+
+        then:
+        result.getOutput().contains("Version: 1.0.0-0-g${shortCommit}")
+
+        where:
+        config << configs(gitDescribeArgs: "--long")
+    }
+
+    def "semverGit {} block overrides project.ext settings"() {
+        when:
+        setupPluginWithClasspath(config)
+        init()
+        commit()
+        tag("new-version-1.0.0")
+        commit()
+        touchBuildFile()
+        BuildResult result = gradleBuild("showVersion")
+
+        then:
+        result.getOutput().contains("Version: 2.0.0-new_-NEW_DIRTYsnapshot")
+
+        where:
+        config = """
```

Comment:
I like this clever use of where, unfortunately this also affects the test name giving this test the name: `semverGit {} block overrides project.ext settings [config: plugins { id "com.cinnober.gradle.semver-git" apply false } ext.nextVersion = 'minor' ext.snapshotSuffix = 'LEGACYSNAPSHOT' ext.dirtyMarker = '-legacydirty' ext.gitDescribeArgs = '--match *[0-9].abc.[0-9]*.def.[0-9]*' ext.semverPrefix = 'legacy-version-' apply plugin: 'com.cinnober.gradle.semver-git' semverGit { nextVersion = 'major' snapshotSuffix = 'new_<dirty>snapshot' dirtyMarker = '-NEW_DIRTY' gitDescribeArgs = '--match *[0-9].[0-9]*.[0-9]* --long' prefix = 'new-version-' } , #0]
` in tools and reports 😀

Reply:
Yes, I noticed that too :laughing: I experimented with a few alternatives, none of which really turned out any better:

- Parameterize over the config `Map` instead of the expanded config file contents
  - This also requires an additional parameter to select which kind of config to generate. It can be embedded into the same `configs()` return value and `setupPluginWithClasspath` can "dynamic dispatch" on it, but the resulting test names aren't that much better. Doesn't seem worth the increased source code noise.

- Parameterize over `configType << ["legacy", "extension"]` instead of the config `Map`.
  - This works well enough, but then the `configType` argument (an invariant across all of the tests) gets mingled in with the config `Map` argument (varies between tests), which entangles two orthogonal concerns. Doesn't seem worth the reduced code clarity.

- Parameterize over the `setupPluginWithClasspath` function itself instead of its argument.
  - This allows the config `Map` to stay in the `when:` block instead (nice!), but makes the test names entirely incomprehensible (they show a closure type instead of the config contents).